### PR TITLE
Falcon40B model updates for prefill/decode on 4 and 8 chips

### PR DESCRIPTION
**Pull request by TT-BrianLiu**

### CODE REVIEW

**Review by boris-drazic — APPROVED** [review]

File: conftest.py [hunk 1]
```diff
@@ -295,6 +295,23 @@ def pcie_devices(request):
...
 
 
+@pytest.fixture(scope="function")
+def all_devices(request):
+    import tt_lib as ttl
+
+    num_devices = ttl.device.GetNumAvailableDevices()
+
+    # Get only physical devices
+    devices = ttl.device.CreateDevices([i for i in range(num_devices)])
```
Comment (tt-rkim) [inline conftest.py:305]: How about `list(range(num_devices))`?

Imagine a DevInfra person giving out code advice? Unthinkable.

File: conftest.py [hunk 2]
```diff
@@ -295,6 +295,23 @@ def pcie_devices(request):
...
+
+    yield [devices[i] for i in range(num_devices)]
```
Comment (tt-rkim) [inline conftest.py:307]: Could this also just be `yield list(devices)`?

File: models/utility_functions.py [hunk 3]
```diff
@@ -956,6 +956,24 @@ def skip_for_grayskull(reason_str="not working for Grayskull"):
...
+    First 4 devices are PCIE devices so we can just extract and return.
+    For 8 devices, return in a ring pattern.
+    """
+    assert num_devices <= len(all_devices), "Not enough devices detected!"
+
+    if num_devices <= 4:
+        return all_devices[:num_devices]
+    elif num_devices == 8:
+        # TODO: Generalize this for different arch
+        hamiltonian_ring_indices = [0, 7, 6, 1, 2, 5, 4, 3]
```
Comment (tt-rkim) [inline models/utility_functions.py:971]: Jeez hamiltonian eh?

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_all_gather.py [hunk 4]
```diff
@@ -38,17 +49,35 @@
         ttl.tensor.MemoryConfig(buffer_type=ttl.tensor.BufferType.L1),
     ],
 )
-@pytest.mark.parametrize("num_links", [1, 2])
 def test_all_gather(
-    pcie_devices, input_shape, dim, num_links, layout, mem_config, use_program_cache, function_level_defaults
+    all_devices,
```
Comment (tt-rkim) [inline tests/tt_eager/python_api_testing/unit_testing/misc/test_all_gather.py:53]: Did you get a chance to stress this on various num_device configs?

**Review by tt-rkim — APPROVED** [review]
Looks good, just a couple of questions.

**Review by arakhmati — APPROVED** [review]

File: tests/tt_eager/python_api_testing/unit_testing/misc/test_all_gather.py [hunk 5]
```diff
@@ -38,17 +49,38 @@
...
+    input_shape,
+    dim,
+    num_links,
+    layout,
+    mem_config,
+    use_program_cache,
+    function_level_defaults,
 ):
+    # TODO: Switch to all_devices once we figure out what works on pipelines
+    devices = pcie_devices
```
Comment (cfjchu) [inline tests/tt_eager/python_api_testing/unit_testing/misc/test_all_gather.py:64]: can we change this to `all_devices`?

**Pull request merged**